### DISCUSSION TITLE
Eheim API Feature Requests

### DISCUSSION THREAD

**Discussion by Lichtenfeld** [davidm-glitch/home-assistant-eheim-digital#21] [Ideas]
Dear Dev-Team and Community,
I am a Developer from Eheim we are planing to release a new Implementation and Documentation of an REST API for our Devices.
We found your great Project and would like to collect some Inspirations and ideas what you guys would like to have implemented.

Feel Free to post any Thoughts :wink:

**Comment by davidm-glitch** [comment]
Hello @Lichtenfeld!

Thanks for getting in touch with us!

The best thing would be an official Home Assistant integration from Eheim ;)

Apart from that, it would be nice to transfer clearly defined values. For example, there is a degree of filter contamination in the web interface. However, this is somehow calculated and displayed by the web interface and not transmitted at all via websocket. 

It would also be nice to see the API receive standardized naming. Camel case, snake case and German and English mixed, everything can currently be found :-D

Which interface should the API work with? Again via websocket?

I'm looking forward to seeing what happens next and hope to hear a lot from you ;)

**Reply by Lichtenfeld** [reply]
Oh yeah, an official HA Integration would be good, but for now we like to open up some API endpoints that people can interact with more easily.
The Web socket connection is for the Devices to exchange messages and configuration internally.
For the API interface, we like to open some HTTP REST endpoints to read actual Values and write some Settings or switch them.
This should be more intuitive to use ;)

Perfect, I will keep you updated.

**Comment by Cl3tus** [comment]
@Lichtenfeld wow, opening this discussion is great! I actually just bought myself an Eheim filter, just because i could read the values in Home Assistant thanks to this great repo! The future of implementation in my Home Assitant for these devices depends on the future expansions of my Aquarium with more Eheim products :)

A REST API would be great actually. Especially the writing some settings. I'm currently an owner of the Eheim Professional 5e 600 filter. And i would really benefit from setting the pump speeds for example.

Also, the Websocket seems to loose some connectivity from time to time, i'm unsure if REST will provide a more stable API. that's not my ballgame.

Looking forward to new API's and implementations!

**Reply by davidm-glitch** [reply]
As we are not professional developers there might be a way to make the websocket connection more solid. But as @Lichtenfeld already mentioned above, this is not the intended interface for this kind of application. 

With a Rest API you are sending requests to the device to catch the data, should be much easier for us to implement this.

And with a developer from Eheim we have a good starting point to make this integration work more reliable and for all the Eheim digital products! Cant wait! 👯

**Reply by Cl3tus** [reply]
Can't wait as well. I'm an IT professional with enough knowledge about the IT workfield. Unfortunately i'm not a developer at all. I do my things with Bash and Powershell. But i think i have enough under my belt to be of assistance!

**Comment by davidm-glitch** [comment]
@jkhalil did you see this? 🥳

**Comment by JakeCK** [comment]
Only indirectly related but it would be nice to be able isolate these devices from the cloud without causing them to periodically panic such as when I block them at the router.

**Reply by davidm-glitch** [reply]
Thats a good point. The API should be reachable locally and not only via the cloud!

**Reply by Lichtenfeld** [reply]
Yea, that's what we intend to do, to give each Device some API Endpoints.
To answer your Request @JakeCK, at the moment there is no cloud.

**Reply by davidm-glitch** [reply]
Will it be possible to call each device by a dedicated IP?
I think this would be easier to integrate. Now you just have the master IP and alle the other devices communicate over it

**Reply by jkhalil** [reply]
Hub has devices and IPs to id what devices are set up, use IPs to call devices specific API endpoints to get device specific data?

**Reply by JakeCK** [reply]
> Yea, that's what we intend to do, to give each Device some API Endpoints. To answer your Request @JakeCK, the Devices do not really connect to a cloud, they only contact the update Server to ask if there is any new Version and to send out Service/Device-Information Mails to the User.

Ah. Thats nice to know. I've just noticed when they're blocked from internet access they become unresponsive for a short time periodically. 

With the addition of a local API it would be very nice to have a setting to turn off device status and update  checks. And make it an entirely local device.

**Reply by Lichtenfeld** [reply]
@davidm-glitch that's a bit complicated because the setup needs to be am meshed Configuration in many cases, so we thought about the Master Device as "API Gateway" and you can send and gather requests to the Clients via their MAC addresses through the Master.

@JakeCK ok, we consider this as Option.

**Comment by Cl3tus** [comment]
@Lichtenfeld would it be possible to test custom BIN files (or whatever filestructure you use for Binaries)? (Beta testing?)

**Reply by Lichtenfeld** [reply]
Hey @Cl3tus, that's not planned at the moment, but I can mention that there is interest.

**Comment by jkhalil** [comment]
Yes! This is amazing @Lichtenfeld! I find the weakest part of the integrations the websocket trickery/slight reverse engineering I had to use to get data.

I also started heavily rewriting the integration in preparing to submit it as an official core integration. 

One recommendation/request I have is support for Zeroconf which is essential MDNS. Would make integration into HA more seemless. 😃

**Reply by Lichtenfeld** [reply]
Oh yeah thanx @jkhalil that's something that would make Integration much easier.
I will mark this topic as important.

**Comment by seiferrtalle** [comment]
This is a great new discussion here and sounds promising and positive.  I am not a programmer, but I am also a user of an Eheim professionel 5e 600T, an intensive user of Home Assistant and would be very happy about a stable integration! 😃

@Lichtenfeld Thank you very much for taking the initiative and contacting this community here! 👍

BR,
Ralf

**Comment by Lichtenfeld** [comment]
Hello everyone,

We are excited to announce that a Swagger draft of our API will be available next week.
We are eager to hear your thoughts and feedback.

Stay tuned for more updates, and we look forward to engaging with the community to refine and improve the API.

Best regards and have a nice Weekend,
Your Eheim Dev Team

**Reply by davidm-glitch** [reply]
cant wait!

**Reply by jkhalil** [reply]
Exciting

**Reply by Lichtenfeld** [reply]
**Good morning,**
Unfortunately, we have to postpone the release as we are still waiting for internal approval. We ask for your understanding and will send you the promised documentation as soon as possible.

**Reply by jkhalil** [reply]
No rush - just excited to refactor with official API. 😁

**Reply by JakeCK** [reply]
@Lichtenfeld this is probably not the best place to ask this but can you please pass on message to those who develop the firmware to do something about the obscene number of email notifications that are sent when there is a firmware update. I have been driven to removing my email address to stop getting them. It is seriously anywhere from 4-10 a day and will completely bury an important notification such as temperature issue or low food on the feeder. And I have no desire to update the firmware unless it fixes an issue im experiencing. 

![Screenshot_20240818-031853](https://github.com/user-attachments/assets/e673fc2f-0b0e-4710-a669-b7309e3e1e83)

**Reply by Lichtenfeld** [reply]
Hello @JakeCK, thank you for this Information!That looks like a mess, i'll pass this to our Devs to find a suitable solution that it get send out once, or you can somehow control them.
For requests like this you can normally write the Support team, they relay that to the right team. (eheim.service@eheim.com)
I would recommend updating to the last Version, there were a lot of mayor fixes and stability/ speed improvements.

**Reply by JakeCK** [reply]
> Hello @JakeCK, thank you for this Information!That looks like a mess, i'll pass this to our Devs to find a suitable solution that it get send out once, or you can somehow control them. For requests like this you can normally write the Support team, they relay that to the right team. ([eheim.service@eheim.com](mailto:eheim.service@eheim.com)) I would recommend updating to the last Version, there were a lot of mayor fixes and stability/ speed improvements.

Is there a change log of the latest firmware?

**Comment by mauriceatkinson** [comment]
hi @Lichtenfeld any update on when your code will be released? cheers

**Reply by Lichtenfeld** [reply]
Hello @mauriceatkinson the Implementation is almost finished, we're currently working on the Software Tests, writing the Documentation and fixing some issues. I hope we'll be able to release it in two to three weeks. :+1:

**Reply by mauriceatkinson** [reply]
Fantastic as I am looking to purchase about 50 auto feeders and wanting to use home assistant to automate the feeding times and amounts.

**Reply by davidm-glitch** [reply]
@Lichtenfeld how is the current status?

**Reply by mauriceatkinson** [reply]
hi @Lichtenfeld any update on your release?  I am currently using HA and Selenium to automate the feeding of my test autofeeder+ unit but I need to scale it to 150 units so I am hoping your code will be better than what I have!!

**Comment by Stiesel** [comment]
Es gab zumindest ein Software-Update, das ich heute Morgen auf der Eheim-Website festgestellt habe. Das Update ist jetzt vom 14. bzw. 16.10.2024.

There was at least one software update I noticed this morning on the Eheim website. The update is now from 14. or 16.10.2024.

**Reply by mauriceatkinson** [reply]
@Stiesel yes I saw that, the install guide gives no details apart from setting up a wifi connection on the autofeeder.  There is no information as to how this is integrated into home assistant.  Hopefully this is comong soon!!

**Reply by Lichtenfeld** [reply]
Good Morning, to give you a little hint, the API is already part of the new Software Update [S1023.W1026 | S2036.W2035]. But we're currently still working on the Documentation Platform and have to get the permit to publish it.
I'll Post an Update as soon we can release it ;)

**Reply by davidm-glitch** [reply]
Just checked the js file real quick, this was the only thing I could find.

`"http://localhost:3000/api/day-cycle?" + a.searchQuery + "=" + a.searchTerm`

Tried to connect to it but didnt work for now. Have to check again later when I have more time :)

**Reply by davidm-glitch** [reply]
Is there a cloud connection incoming? ;-)

`
establishCloudConnection: function() {
                var t = "https://aqualed.herokuapp.com/device/aquaria/" + this.mac + ".json?uid=" + this.uid;
                printLError("establishCloudConnection: " + t);
                var n = this;
                e({
                    url: t,
                    method: "GET",
                    cache: !1
                }).then(function(e) {
                    printLLog("Connection Successful"),
                    printLLog(e),
                    n.connectedCloud = !0,
                    n.stopReloadFunctions()
                }).catch(function(e) {
                    printLLog("Error Callback"),
                    printLLog(e),
                    setTimeout(function() {
                        n.reconnect()
                    }, n.reconnectIntervalRestart)
                })
            }
`

**Reply by davidm-glitch** [reply]
currently trying to check the firmware file itself. 

It is not very readable but this line catched my attention

`http:// /login  /demoUse    /kindle-wifi/wifistub.html  /serverLog  /sys    /pH4    /pH7    /updateFile /getFirmware    /add    text/plain  Operating Minutes:  mailTest    /formatspiffs   /formateeprom   /update /getUpdateStatus    /usb    /preset /setVersion /pwm    /fota   /heap   /api/`

When trying to open /api or /login I get a browser login window. But what might be the credentials?

/serverLog and /sys and /heap are working. But not really usefull for our API.

**Comment by joshua-dessol** [comment]
The API credentials are the following and can be extracted from the websocket:

`"api_usrName": "api",
 "api_password": "admin",`
 
 I'm able to connect through this but I'm struggling to figure out the correct URL parameters. Might dig around some more tomorrow but if anyone else might give it a shot it would be of great help.

**Comment by madxdev** [comment]
It seems to be functional. Randomly, I found and tested the following.

GET
http://MASTER_DEVICE_IP/api/devicelist
http://MASTER_DEVICE_IP/api/userdata

http://PHCONTROL_DEVICE_IP/api/phcontrol
http://FILTER_DEVICE_IP/api/professionel5e

POST 
http://PHCONTROL_DEVICE_IP/api/phcontrol/manual 
(Body raw JSON)
BODY
{
    "title": "SET_PH_PARAM",
    "to": "48:E7:29:2A:0D:A5",
    "sollPH": 69,
    "active": 1,
}

**Reply by davidm-glitch** [reply]
@jkhalil can you already start with creating a new branch for the API? :)

**Reply by madxdev** [reply]
Additional potential API endpoints suitable for exploration.

/phcontrol
/phcontrol/active
/phcontrol/manual
/phcontrol/bio

/professionel5e
/professionel5e/pulse
/professionel5e/bio
/professionel5e/constant
/professionel5e/manual
/professionel5e/active

/autofeeder
/autofeeder/config
/autofeeder/bio
/autofeeder/feed
/autofeeder/active
/autofeeder/full

/thermocontrol
/thermocontrol/config
/thermocontrol/bio
/thermocontrol/manual
/thermocontrol/active

/climacontrol
/climacontrol/config
/climacontrol/bio
/climacontrol/manual
/climacontrol/active

/light
/light/daycycle
/light/manual
/light/daycycle/acclimatisation
/light/daycycle/clouds
/light/daycycle/moon
/light/daycycle/multicollor
/light/daycycle/config

/reeflexuv
/reeflexuv/active
/reeflexuv/daycycle
/reeflexuv/pause
/reeflexuv/boost

/classicvario
/classicvario/active
/classicvario/manual
/classicvario/bio
/classicvario/pulse

**Reply by Lichtenfeld** [reply]
Wow, guys, you did a great job extracting some endpoints from the binary!
The reason that there was no Update is, that we are currently very busy with some major fixes and additions.
Because you are willing to move on, we just said that we'll give you the access to the beta documentation in the next few days.

/[devicename]  are GET Endpoints and the /[devicename]/[action] are POST Endpoints to set the parameters.

Parameter Checks are partly included at the moment, so you still get a success message, but the values are mapped to the max-min range.

**Reply by jkhalil** [reply]
Sure - I have a branch with some modernization updates I was 80% done with. When I get the documentation, I can create a new branch and try to move over the API calls.

**Reply by davidm-glitch** [reply]
@Lichtenfeld hows it going? :)

**Comment by madxdev** [comment]
Another thing that comes to mind is the need to filter the range of input values for entities. For instance, I discovered that the API allows setting dangerous combinations. For example, when KH is 8, it also allows a sollPH of 66, which is extremely dangerous for animals. The web interface, however, has safeguards in place and does not permit this.

**Comment by Lichtenfeld** [comment]
Dear Community,

We are excited to finally release the long-awaited API Documentation! We truly appreciate your patience and support.

You can now find the full documentation here: ** [api.eheimdigital.com](https://api.eheimdigital.com/) **. This includes:
✅ Comprehensive endpoints overview
✅ Usage examples
✅ Authentication details

### Important Notes:

🔹 The API is available starting from software versions S2036 and S1024 from (release: 17.12.2024).
🔹 Please note that the following devices currently have an issue when acting as a master – they are unable to process POST requests. This will be fixed in the next release:
- Professional5e
- climacontrol+
- thermocontrol+e
- LEDControl+
- autofeeder+

We hope this makes working with the API easier and more efficient for everyone. If you have any feedback, suggestions, or spot any issues, feel free to write a comment!

Happy coding!
Your Eheim Dev Team

![grafik](https://github.com/user-attachments/assets/d2193eed-4f28-4fd5-bbef-f2a8f6d45e4d)
